### PR TITLE
[WIP] Filter interface

### LINKED ISSUE #168: Filtering interface / endpoints
Support interactive filtering of datasets via mongo / csv query endpoint(s)

### PR DESCRIPTION
**Pull request by alex-r-bigelow**
Adds filtering capabilities to Resonant Laboratory

Fixes #168, #284

### CODE REVIEW

**Comment by alex-r-bigelow** [comment]
At this point in time, this PR includes all the client-side basics of the filter interface, with histogram calculations function correctly for mongo-based datasets.

The remaining functionality (mapreduce for non-mongo datasets, pulling the correct data in the table) depends on a download endpoint that accepts a filter parameter, i.e. we are waiting on [this](https://github.com/OpenGeoscience/girder_db_items/pull/8) PR to be merged.

I will start a separate `filter-interface-download` branch that will refactor resonant lab to make use of the new girder_db_items plugin, and merge it into this PR once girder_db_items is ready.

File: app/resonant-laboratory/grunt-watch-plugin.js [hunk 1]
```diff
@@ -12,7 +12,7 @@ module.exports = {
   apply: function (compiler) {
     compiler.plugin('emit', (compilation, callback) => {
       if (compilation.options.watch === true) {
-        exec('cd $GIRDER_PATH && grunt', function (error, stdout, stderr) {
+        exec('cd $GIRDER_PATH && grunt --force', function (error, stdout, stderr) {
```
Comment (ronichoudhury) [inline app/resonant-laboratory/grunt-watch-plugin.js]: What's the reason for adding `--force` here?

**Comment by alex-r-bigelow** [comment]
At this point, this branch requires two (possibly three) major changes before it's eligible to be merged into master:
- [ ] We need `semantic_access()` to support DB assetstores—mongo-based datasets will show a broken table view / connected visualizations will not work until this is fixed
   - [ ] Probably would use the same logic as streaming JSON files?
- [ ] Need support for type casting in queries (#307)—csv-based datasets will have odd/incorrect filter behavior until this is fixed

@ronichoudhury, I'm happy to work on either of these; I just want to make sure I won't be stepping on your toes if I proceed.

File: app/resonant-laboratory/grunt-watch-plugin.js (hunk 1 above)
Comment (alex-r-bigelow) [inline app/resonant-laboratory/grunt-watch-plugin.js]: For some reason grunt is tripping up (and crashing) on the `database_assetstore` plugin. It's possible there's something weird in my girder installation as a result of switching branches... I'll try a fresh girder installation and see if this problem goes away.

File: package.json [hunk 2]
```diff
@@ -9,7 +9,7 @@
     "build:laboratory": "cd app/resonant-laboratory && webpack",
     "build:laboratory:tests": "webpack --config ./app/resonant-laboratory/webpack-testing.config.js",
     "test:laboratory:unit": "npm run test:laboratory:unit:raw | tap-spec",
-    "test:laboratory:unit:raw": "node build/resonant-laboratory/tests.js",
+    "test:laboratory:unit:raw": "for t in `ls build/resonant-laboratory/*.tests.js`; do node $t; done",
```
Comment (ronichoudhury) [inline package.json]: What was the reason for this change?

Comment (alex-r-bigelow) [inline package.json]: I added a few of my own tests; if there's a better way to do this, I'm happy to change it.

Comment (ronichoudhury) [inline package.json]: Oh I see. I think you can pass multiple files directly to `node` but maybe that's not right.

This isn't that big of a deal, it's just that it depends on a unix-type shell to work properly.

File: app/resonant-laboratory/grunt-watch-plugin.js (hunk 1 above)
Comment (ronichoudhury) [inline app/resonant-laboratory/grunt-watch-plugin.js]: In that case I'd prefer we not just put `--force` in here. Let's try to figure out what's going wrong on your machine first.

**Comment by alex-r-bigelow** [comment]
To replicate both behaviors:

1) Run the new `scripts/populateGirder.py` script—this will upload a bunch of flat CSV files and mongo db files

2) Create a new project with the `cars` dataset (this one is a CSV file)
  - The filters should misbehave a bit—you can inspect the filters that the interface is generating by inspecting the `getHistograms` requests. Or you can see the human-friendly version of the filters by typing (Chrome will probably offer to autocomplete the right girder ID):

    `mainPage.loadedDatasets[' --- the dataset's girder ID --- '].listAllFilterExpressions()`

    or to see the resulting AST tree:

    `mainPage.loadedDatasets[' --- the dataset's girder ID --- '].getFilterAstTree()`

3) Switch to the `airports` dataset (this one is a mongo collection)
  - Here the filters should work as expected
  - The table view should be broken; if you inspect the `download?extraParameters=...` request, you'll see the kind of thing that comes back from the current download endpoint. This *should* look a little different:
    - The download endpoint needs to use the filter / limit / offset parameters that we're sending it
    - We should be getting the contents of the `data` property without the meta information (not a big deal; we can always do a check and pull it out if that's what's supposed to come back from downloading a database table)

**Comment by alex-r-bigelow** [comment]
I accidentally closed this PR... because of difficulties in reopening it, we have moved it to #316

**Pull request closed without merging**